### PR TITLE
[DF] [ROOT-9461] Introduced Kahan Sum as a custom action

### PR DESCRIPTION
**Pull request by imaxoi**
Introduced tutorial on how to create a Kahan summation, with a simple example on how it can give better results than the classical one.

This issue solves the Jira issue ROOT-9461

### CODE REVIEW

File: tutorials/dataframe/df022_createKahan.C [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+/// \ingroup tutorial_dataframe
+/// \notebook
+/// This tutorial shows how to implement a Kahan summation custom action.
+///
+/// \macro_code
+///
+/// \date July 2018
+/// \author Enrico Guiraud, Danilo Piparo, Massimo Tumolo
+
+class KahanSum : public ROOT::Detail::RDF::RActionImpl<class KahanSum> {
```
Comment (dpiparo) [inline tutorials/dataframe/df022_createKahan.C]: I would template on the FP type (float or double) and propagate accordingly to the class. For example, I would also add a static assert that tells the user if she requested this summation on integer types.
See http://www.cplusplus.com/reference/type_traits/is_floating_point.

File: tutorials/dataframe/df022_createKahan.C [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+   std::shared_ptr<Result_t> GetResultPtr() const { return fResultSum; }
+
+   void Initialize() {}
+   void InitTask(TTreeReader *, unsigned int) {}
+
+   template <typename T>
+   void Exec(unsigned int slot, T x)
+   {
+      double y = x - fCompensations[slot];
```
Comment (dpiparo) [inline tutorials/dataframe/df022_createKahan.C]: Would it make sense to extract the Kahan summation algo in a private method and use it both here and in Finalize?

File: tutorials/dataframe/df022_createKahan.C [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+void df022_createKahan()
+{
+   // We enable implicit parallelism
+   ROOT::EnableImplicitMT(8);
+
+   std::vector<double> source(10);
+   double standardSum = 0;
+   double ele = 0;
+   for (int i = 1; i <= 10; ++i) {
+      ele = (i % 2 == 0) ? ((double)1) / 100000000 : ((double)1) * 100000000;
```
Comment (dpiparo) [inline tutorials/dataframe/df022_createKahan.C]: Could maybe `((double)1) / 100000000 : ((double)1) * 100000000` be something like 00000000.1 : 100000000;

File: tutorials/dataframe/df022_createKahan.C [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+void df022_createKahan()
+{
+   // We enable implicit parallelism
+   ROOT::EnableImplicitMT(8);
+
+   std::vector<double> source(10);
+   double standardSum = 0;
+   double ele = 0;
+   for (int i = 1; i <= 10; ++i) {
```
Comment (dpiparo) [inline tutorials/dataframe/df022_createKahan.C]: maybe start from 0?

File: tutorials/dataframe/df022_createKahan.C [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+      ele = (i % 2 == 0) ? ((double)1) / 100000000 : ((double)1) * 100000000;
+      source[i - 1] = ele;
+      standardSum += ele;
+   }
+
+   ROOT::RDataFrame d(10);
+   auto dd = d.DefineSlotEntry("x1", [&source](unsigned int slot, double entry) {
```
Comment (dpiparo) [inline tutorials/dataframe/df022_createKahan.C]: The type of entry should be ULong64_t, right?

File: tutorials/dataframe/df022_createKahan.C [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      (void)slot;
+      return source[entry];
+   });
+
+   using Helper_t = KahanSum;
```
Comment (dpiparo) [inline tutorials/dataframe/df022_createKahan.C]: Isn't this useless?

File: tutorials/dataframe/df022_createKahan.C [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+         sum = t;
+      }
+      *fResultSum = sum;
+   }
+};
+
+void df022_createKahan()
+{
+   // We enable implicit parallelism
+   ROOT::EnableImplicitMT(8);
```
Comment (dpiparo) [inline tutorials/dataframe/df022_createKahan.C]: Maybe one could put here a 2 or a 4?

File: tutorials/dataframe/df022_createKahan.C (hunk 1 above)
Comment (dpiparo) [inline tutorials/dataframe/df022_createKahan.C]: I would use the final keyword.

File: tutorials/dataframe/df022_createKahan.C [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+   void Exec(unsigned int slot, const T &vs)
+   {
+      for (auto &&v : vs) {
+         Exec(slot, v);
+      }
+   }
+
+   void Finalize()
+   {
+      double sum = 0;
```
Comment (dpiparo) [inline tutorials/dataframe/df022_createKahan.C]: this should be initialised by T(0)

**Comment by imaxoi** [comment]
@dpiparo Thanks for the review, fixed.

**Comment by dpiparo** [comment]
@imaxoi let's wait for the builds to be green.

**Comment by imaxoi** [comment]
@phsft-bot build, and don't timeout

File: tutorials/dataframe/df022_useKahan.C [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+      for (int i = 0; i < fNSlots; ++i) {
+         KahanAlgorithm(fPartialSums[i], sum, compensation);
+      }
+      *fResultSum = sum;
+   }
+};
+
+void df022_useKahan()
+{
+   static const int NR_ELEMENTS = 20;
```
[the remark this answers is not in the document]
Comment (imaxoi) [reply tutorials/dataframe/df022_useKahan.C]: Yes, done

**Pull request merged**